File: server/tasks/ifs.py
#!/usr/bin/env python3
# ifs.py

"""
    Copyright (C) 2012-2016 Rudolf Cardinal (rudolf@pobox.com).
    Department of Psychiatry, University of Cambridge.
    Funded by the Wellcome Trust.

    This file is part of CamCOPS.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
"""

from ..cc_modules.cc_constants import (
    CTV_DICTLIST_INCOMPLETE,
    DATA_COLLECTION_UNLESS_UPGRADED_DIV,
    INVALID_VALUE,
    PV,
)
from ..cc_modules.cc_html import (
    answer,
    get_correct_incorrect_none,
    td,
    tr,
    tr_qa,
)
# from ..cc_modules.cc_string import WSTRING
from ..cc_modules.cc_task import Task


# =============================================================================
# IFS
# =============================================================================

class Ifs(Task):
    Q4_DIGIT_LENGTHS = list(range(2, 7 + 1))
    Q6_SEQUENCE_NUMS = list(range(1, 4 + 1))
    Q7_PROVERB_NUMS = list(range(1, 3 + 1))
    Q8_SENTENCE_NUMS = list(range(1, 3 + 1))
    SIMPLE_Q = (
        ["q1", "q2", "q3", "q5"] +
        ["q6_seq{}".format(n) for n in Q6_SEQUENCE_NUMS] +
        ["q7_proverb{}".format(n) for n in Q7_PROVERB_NUMS] +
        ["q8_sentence{}".format(n) for n in Q8_SENTENCE_NUMS]
    )

    tablename = "ifs"
    shortname = "IFS"
    longname = "INECO Frontal Screening"
    fieldspecs = [
        dict(name="q1", cctype="INT", pv=[0, 1, 2, 3],
             comment="Q1. Motor series (motor programming)"),
        dict(name="q2", cctype="INT", pv=[0, 1, 2, 3],
             comment="Q2. Conflicting instructions "
                     "(interference sensitivity)"),
        dict(name="q3", cctype="INT", pv=[0, 1, 2, 3],
             comment="Q3. Go/no-go (inhibitory control)"),
    ]
    for seqlen in Q4_DIGIT_LENGTHS:
        fieldspecs.extend([
            dict(
                name="q4_len{}_1".format(seqlen), cctype="BOOL", pv=PV.BIT,
                comment="Q4. Digits backward, length {}, trial 1".format(
                    seqlen)),
            dict(
                name="q4_len{}_2".format(seqlen), cctype="BOOL", pv=PV.BIT,
                comment="Q4. Digits backward, length {}, trial 2".format(
                    seqlen)),
        ])
    fieldspecs.extend([
        dict(name="q5", cctype="INT", pv=[0, 1, 2],
             comment="Q5. Verbal working memory"),
    ])
    for n in Q6_SEQUENCE_NUMS:
        fieldspecs.extend([
            dict(
                name="q6_seq{}".format(n), cctype="INT", pv=[0, 1],
                comment="Q6. Spatial working memory, sequence {}".format(n)),
        ])
    for n in Q7_PROVERB_NUMS:
        fieldspecs.extend([
            dict(name="q7_proverb{}".format(n), cctype="FLOAT",
                 min=0, max=1,
                 comment="Q7. Proverb {} (1 = correct explanation, "
                         "0.5 = example, 0 = neither)".format(n)),
        ])
    for n in Q8_SENTENCE_NUMS:
        fieldspecs.extend([
            dict(name="q8_sentence{}".format(n), cctype="INT", pv=[0, 1, 2],
                 comment="Q8. Hayling, sentence {}".format(n)),
        ])
    has_clinician = True

    def get_trackers(self):
        scoredict = self.get_score()
        return [
            {
                "value": scoredict['total'],
                "plot_label": "IFS total score (higher is better)",
                "axis_label": "Total score (out of 30)",
                "axis_min": -0.5,
                "axis_max": 30.5,
            },
            {
                "value": scoredict['wm'],
                "plot_label": "IFS working memory index (higher is better)",
                "axis_label": "Total score (out of 10)",
                "axis_min": -0.5,
                "axis_max": 10.5,
            },
        ]

    def get_summaries(self):
        scoredict = self.get_score()
        return [
            self.is_complete_summary_field(),
            dict(name="total", cctype="FLOAT",
                 value=scoredict['total'],
                 comment="Total (out of 30, higher better)"),
            dict(name="wm", cctype="INT",
                 value=scoredict['wm'],
                 comment="Working memory index (out of 10; sum of Q4 + Q6"),
        ]

    def get_clinical_text(self):
        scoredict = self.get_score()
        if not self.is_complete():
            return CTV_DICTLIST_INCOMPLETE
        return [
            {
                "content": "Total: {t}/30; working memory index {w}/10".format(
                    t=scoredict['total'],
                    w=scoredict['wm'],
                )
            }
        ]

    def get_score(self):
        q1 = getattr(self, "q1", 0) or 0
        q2 = getattr(self, "q2", 0) or 0
        q3 = getattr(self, "q3", 0) or 0
        q4 = 0
        for seqlen in self.Q4_DIGIT_LENGTHS:
            val1 = getattr(self, "q4_len{}_1".format(seqlen))
            val2 = getattr(self, "q4_len{}_2".format(seqlen))
            if val1 or val2:
                q4 += 1
            if not val1 and not val2:
                break
        q5 = getattr(self, "q5", 0) or 0
        q6 = self.sum_fields(["q6_seq" + str(s) for s in range(1, 4 + 1)])
        q7 = self.sum_fields(["q7_proverb" + str(s) for s in range(1, 3 + 1)])
        q8 = self.sum_fields(["q8_sentence" + str(s) for s in range(1, 3 + 1)])
        total = q1 + q2 + q3 + q4 + q5 + q6 + q7 + q8
        wm = q4 + q6  # working memory index (though not verbal)
        return dict(
            total=total,
            wm=wm
        )

    def is_complete(self):
        if not self.field_contents_valid():
            return False
        if not self.are_all_fields_complete(self.SIMPLE_Q):
            return False
        for seqlen in self.Q4_DIGIT_LENGTHS:
            val1 = getattr(self, "q4_len{}_1".format(seqlen))
            val2 = getattr(self, "q4_len{}_2".format(seqlen))
            if val1 is None or val2 is None:
                return False
            if not val1 and not val2:
                return True  # all done
        return True

    def get_simple_tr_qa(self, qprefix):
        q = self.WXSTRING(qprefix + "_title")
        val = getattr(self, qprefix)
        if val is not None:
            a = self.WXSTRING(qprefix + "_a" + str(val))
        else:
            a = None
        return tr_qa(q, a)

    def get_task_html(self):
        scoredict = self.get_score()
        h = """
            <div class="summary">
                <table class="summary">
                    {complete_tr}
                    <tr>
                        <td>Total (higher better)</td>
                        <td>{total} / 30</td>
                    </td>
                    <tr>
                        <td>Working memory index <sup>1</sup></td>
                        <td>{wm} / 10</td>
                    </td>
                </table>
            </div>
            <table class="taskdetail">
                <tr>
                    <th width="50%">Question</th>
                    <th width="50%">Answer</th>
                </tr>
        """.format(
            complete_tr=self.get_is_complete_tr(),
            total=answer(scoredict['total']),
            wm=answer(scoredict['wm']),
        )

        # Q1
        h += self.get_simple_tr_qa("q1")
        # Q2
        h += self.get_simple_tr_qa("q2")
        # Q3
        h += self.get_simple_tr_qa("q3")
        # Q4
        h += tr(td(self.WXSTRING("q4_title")), td("", td_class="subheading"),
                literal=True)
        required = True
        for n in self.Q4_DIGIT_LENGTHS:
            val1 = getattr(self, "q4_len{}_1".format(n))
            val2 = getattr(self, "q4_len{}_2".format(n))
            q = (
                "… " +
                self.WXSTRING("q4_seq_len{}_1".format(n)) +
                " / " + self.WXSTRING("q4_seq_len{}_2".format(n))
            )
            if required:
                score = 1 if val1 or val2 else 0
                a = (
                    answer(get_correct_incorrect_none(val1)) +
                    " / " + answer(get_correct_incorrect_none(val2)) +
                    " (scores {})".format(score)
                )
            else:
                a = ""
            h += tr(q, a)
            if not val1 and not val2:
                required = False
        # Q5
        h += self.get_simple_tr_qa("q5")
        # Q6
        h += tr(td(self.WXSTRING("q6_title")), td("", td_class="subheading"),
                literal=True)
        for n in self.Q6_SEQUENCE_NUMS:
            nstr = str(n)
            val = getattr(self, "q6_seq" + nstr)
            h += tr_qa("… " + self.WXSTRING("q6_seq" + nstr), val)
        # Q7
        q7map = {
            None: None,
            1: self.WXSTRING("q7_a_1"),
            0.5: self.WXSTRING("q7_a_half"),
            0: self.WXSTRING("q7_a_0"),
        }
        h += tr(td(self.WXSTRING("q7_title")), td("", td_class="subheading"),
                literal=True)
        for n in self.Q7_PROVERB_NUMS:
            nstr = str(n)
            val = getattr(self, "q7_proverb" + nstr)
            a = q7map.get(val, INVALID_VALUE)
            h += tr_qa("… " + self.WXSTRING("q7_proverb" + nstr), a)
        # Q8
        q8map = {
            None: None,
            2: self.WXSTRING("q8_a2"),
            1: self.WXSTRING("q8_a1"),
            0: self.WXSTRING("q8_a0"),
        }
        h += tr(td(self.WXSTRING("q8_title")), td("", td_class="subheading"),
                literal=True)
        for n in self.Q8_SENTENCE_NUMS:
            nstr = str(n)
            val = getattr(self, "q8_sentence" + nstr)
            a = q8map.get(val, INVALID_VALUE)
            h += tr_qa("… " + self.WXSTRING("q8_sentence_" + nstr), a)

        h += """
            </table>
            <div class="footnotes">
                [1] Sum of scores for Q4 + Q6.
            </div>
        """ + DATA_COLLECTION_UNLESS_UPGRADED_DIV
        return h

File: tablet/Resources/task/Ifs.js
// Ifs.js

/*
    Copyright (C) 2012-2016 Rudolf Cardinal (rudolf@pobox.com).
    Department of Psychiatry, University of Cambridge.
    Funded by the Wellcome Trust.

    This file is part of CamCOPS.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

/*jslint node: true, newcap: true, nomen: true, plusplus: true, unparam: true */
"use strict";
/*global Titanium */

var DBCONSTANTS = require('common/DBCONSTANTS'),
    dbcommon = require('lib/dbcommon'),
    taskcommon = require('lib/taskcommon'),
    lang = require('lib/lang'),
    tablename = "ifs",
    IMAGE_SWM = '/images/ifs/swm.png',
    DIGIT_ELEMENT_PREFIX = "digit_element_",
    Q4_PAGETAG = "q4page",
    fieldlist = dbcommon.standardTaskFields();

fieldlist.push.apply(fieldlist, dbcommon.CLINICIAN_FIELDSPECS); // Clinician info 1/3
fieldlist.push(
    {name: "q1", type: DBCONSTANTS.TYPE_INTEGER},
    {name: "q2", type: DBCONSTANTS.TYPE_INTEGER},
    {name: "q3", type: DBCONSTANTS.TYPE_INTEGER},
    {name: "q4_len2_1", type: DBCONSTANTS.TYPE_BOOLEAN},
    {name: "q4_len2_2", type: DBCONSTANTS.TYPE_BOOLEAN},
    {name: "q4_len3_1", type: DBCONSTANTS.TYPE_BOOLEAN},
    {name: "q4_len3_2", type: DBCONSTANTS.TYPE_BOOLEAN},
    {name: "q4_len4_1", type: DBCONSTANTS.TYPE_BOOLEAN},
    {name: "q4_len4_2", type: DBCONSTANTS.TYPE_BOOLEAN},
    {name: "q4_len5_1", type: DBCONSTANTS.TYPE_BOOLEAN},
    {name: "q4_len5_2", type: DBCONSTANTS.TYPE_BOOLEAN},
    {name: "q4_len6_1", type: DBCONSTANTS.TYPE_BOOLEAN},
    {name: "q4_len6_2", type: DBCONSTANTS.TYPE_BOOLEAN},
    {name: "q4_len7_1", type: DBCONSTANTS.TYPE_BOOLEAN},
    {name: "q4_len7_2", type: DBCONSTANTS.TYPE_BOOLEAN},
    {name: "q5", type: DBCONSTANTS.TYPE_INTEGER},
    {name: "q6_seq1", type: DBCONSTANTS.TYPE_INTEGER},
    {name: "q6_seq2", type: DBCONSTANTS.TYPE_INTEGER},
    {name: "q6_seq3", type: DBCONSTANTS.TYPE_INTEGER},
    {name: "q6_seq4", type: DBCONSTANTS.TYPE_INTEGER},
    {name: "q7_proverb1", type: DBCONSTANTS.TYPE_REAL},  // can score 0.5
    {name: "q7_proverb2", type: DBCONSTANTS.TYPE_REAL},
    {name: "q7_proverb3", type: DBCONSTANTS.TYPE_REAL},
    {name: "q8_sentence1", type: DBCONSTANTS.TYPE_INTEGER},
    {name: "q8_sentence2", type: DBCONSTANTS.TYPE_INTEGER},
    {name: "q8_sentence3", type: DBCONSTANTS.TYPE_INTEGER}
);

// CREATE THE TABLE

dbcommon.createTable(tablename, fieldlist);

// TASK

function Ifs(patient_id) {
    taskcommon.BaseTask.call(this, patient_id); // call base constructor
}

lang.inheritPrototype(Ifs, taskcommon.BaseTask);
lang.extendPrototype(Ifs, {

    // KEY DATABASE FIELDS (USED BY DatabaseObject)

    _objecttype: Ifs,
    _tablename: tablename,
    _fieldlist: fieldlist,

    // TASK CLASS FIELD OVERRIDES (USED BY BaseTask)

    _extrastringTaskname: "ifs",
    isTaskCrippled: function () {
        return !this.extraStringsPresent();
    },

    // OTHER

    getScore: function () {
        var total = 0,
            wm = 0,
            q1,
            q2,
            q3,
            q4,
            q5,
            q6,
            q7,
            q8,
            seqlen,
            val1,
            val2;
        q1 = this.q1 || 0;
        q2 = this.q2 || 0;
        q3 = this.q3 || 0;
        q4 = 0;
        for (seqlen = 2; seqlen <= 7; ++seqlen) {
            val1 = this["q4_len" + seqlen + "_1"];
            val2 = this["q4_len" + seqlen + "_2"];
            q4 += val1 || val2 ? 1 : 0;
            if (!val1 && !val2) {
                break;
            }
        }
        q5 = this.q5 || 0;
        q6 = (
            (this.q6_seq1 || 0) +
            (this.q6_seq2 || 0) +
            (this.q6_seq3 || 0) +
            (this.q6_seq4 || 0)
        );
        q7 = (
            (this.q7_proverb1 || 0) +
            (this.q7_proverb2 || 0) +
            (this.q7_proverb3 || 0)
        );
        q8 = (
            (this.q8_sentence1 || 0) +
            (this.q8_sentence2 || 0) +
            (this.q8_sentence3 || 0)
        );
        total = q1 + q2 + q3 + q4 + q5 + q6 + q7 + q8;
        wm = q4 + q6;  // working memory index (though not verbal)
        return {total: total, wm: wm};
    },

    // Standard task functions
    isComplete: function () {
        var seqlen,
            val1,
            val2,
            simple_qs = [
                "q1", "q2", "q3", "q5",
                "q6_seq1", "q6_seq2", "q6_seq3", "q6_seq4",
                "q7_proverb1", "q7_proverb2", "q7_proverb3",
                "q8_sentence1", "q8_sentence2", "q8_sentence3"
            ];
        // Obligatory stuff
        if (!taskcommon.isComplete(this, simple_qs)) {
            return false;
        }
        // Q4 (digit span), where we can terminate early
        // The sequences come in pairs. The task terminates when the patient
        // gets both items wrong within the pair.
        for (seqlen = 2; seqlen <= 7; ++seqlen) {
            val1 = this["q4_len" + seqlen + "_1"];
            val2 = this["q4_len" + seqlen + "_2"];
            if (val1 === null || val2 === null) {
                return false;
            }
            if (!val1 && !val2) {
                return true;  // all done
            }
        }
        return true;
    },

    getSummary: function () {
        var score = this.getScore();
        return (
            "Total = " + score.total + "/30; " +
            "working memory index = " + score.wm + "/10" +
            this.isCompleteSuffix()
        );
    },

    getDetail: function () {
        return this.getSummary();
    },

    edit: function (readOnly) {
        var self = this,
            KeyValuePair = require('lib/KeyValuePair'),
            Questionnaire = require('questionnaire/Questionnaire'),
            pages,
            questionnaire,
            proverb_options,
            inhibition_options;

        self.setDefaultClinicianVariablesAtFirstUse(readOnly); // Clinician info 2/3
        proverb_options = [
            new KeyValuePair(this.XSTRING("q7_a_1"), 1),
            new KeyValuePair(this.XSTRING("q7_a_half"), 0.5),
            new KeyValuePair(this.XSTRING("q7_a_0"), 0)
        ];
        inhibition_options = [
            new KeyValuePair(this.XSTRING("q8_a2"), 2),
            new KeyValuePair(this.XSTRING("q8_a1"), 1),
            new KeyValuePair(this.XSTRING("q8_a0"), 0)
        ];

        pages = [
            this.getClinicianDetailsPage(), // Clinician info 3/3

            // ----------------------------------------------------------------
            // Q1
            // ----------------------------------------------------------------
            {
                title: this.XSTRING('q1_title'),
                clinician: true,
                elements: [
                    {
                        type: "QuestionText",
                        bold: true,
                        text: this.XSTRING("q1_instruction_1")
                    },
                    {
                        type: "QuestionText",
                        text: this.XSTRING("q1_instruction_2")
                    },
                    {
                        type: "QuestionText",
                        bold: true,
                        text: this.XSTRING("q1_instruction_3")
                    },
                    {
                        type: "QuestionText",
                        text: this.XSTRING("q1_instruction_4")
                    },
                    {
                        type: "QuestionText",
                        bold: true,
                        text: this.XSTRING("q1_instruction_5")
                    },
                    {
                        type: "QuestionMCQ",
                        mandatory: true,
                        horizontal: false,
                        showInstruction: false,
                        field: "q1",
                        options: [
                            new KeyValuePair(this.XSTRING("q1_a3"), 3),
                            new KeyValuePair(this.XSTRING("q1_a2"), 2),
                            new KeyValuePair(this.XSTRING("q1_a1"), 1),
                            new KeyValuePair(this.XSTRING("q1_a0"), 0)
                        ]
                    }
                ]
            },

            // ----------------------------------------------------------------
            // Q2
            // ----------------------------------------------------------------
            {
                title: this.XSTRING('q2_title'),
                clinician: true,
                elements: [
                    {
                        type: "QuestionText",
                        bold: true,
                        text: this.XSTRING("q2_instruction_1")
                    },
                    {
                        type: "QuestionText",
                        text: this.XSTRING("q2_instruction_2")
                    },
                    {
                        type: "QuestionText",
                        bold: true,
                        text: this.XSTRING("q2_instruction_3")
                    },
                    {
                        type: "QuestionText",
                        text: this.XSTRING("q2_instruction_4")
                    },
                    {
                        type: "QuestionText",
                        bold: true,
                        text: this.XSTRING("q2_instruction_5")
                    },
                    {
                        type: "QuestionMCQ",
                        mandatory: true,
                        horizontal: false,
                        showInstruction: false,
                        field: "q2",
                        options: [
                            new KeyValuePair(this.XSTRING("q2_a3"), 3),
                            new KeyValuePair(this.XSTRING("q2_a2"), 2),
                            new KeyValuePair(this.XSTRING("q2_a1"), 1),
                            new KeyValuePair(this.XSTRING("q2_a0"), 0)
                        ]
                    }
                ]
            },

            // ----------------------------------------------------------------
            // Q3
            // ----------------------------------------------------------------
            {
                title: this.XSTRING('q3_title'),
                clinician: true,
                elements: [
                    {
                        type: "QuestionText",
                        bold: true,
                        text: this.XSTRING("q3_instruction_1")
                    },
                    {
                        type: "QuestionText",
                        text: this.XSTRING("q3_instruction_2")
                    },
                    {
                        type: "QuestionText",
                        bold: true,
                        text: this.XSTRING("q3_instruction_3")
                    },
                    {
                        type: "QuestionText",
                        text: this.XSTRING("q3_instruction_4")
                    },
                    {
                        type: "QuestionText",
                        bold: true,
                        text: this.XSTRING("q3_instruction_5")
                    },
                    {
                        type: "QuestionMCQ",
                        mandatory: true,
                        horizontal: false,
                        showInstruction: false,
                        field: "q3",
                        options: [
                            new KeyValuePair(this.XSTRING("q3_a3"), 3),
                            new KeyValuePair(this.XSTRING("q3_a2"), 2),
                            new KeyValuePair(this.XSTRING("q3_a1"), 1),
                            new KeyValuePair(this.XSTRING("q3_a0"), 0)
                        ]
                    }
                ]
            },

            // ----------------------------------------------------------------
            // Q4
            // ----------------------------------------------------------------
            {
                onTheFly: true,
                pageTag: Q4_PAGETAG
            },

            // ----------------------------------------------------------------
            // Q5
            // ----------------------------------------------------------------
            {
                title: this.XSTRING('q5_title'),
                clinician: true,
                elements: [
                    {
                        type: "QuestionText",
                        bold: true,
                        text: this.XSTRING("q5_instruction_1")
                    },
                    {
                        type: "QuestionText",
                        text: this.XSTRING("q5_instruction_2")
                    },
                    {
                        type: "QuestionText",
                        text: this.XSTRING("q5_instruction_3")
                    },
                    {
                        type: "QuestionMCQ",
                        mandatory: true,
                        horizontal: false,
                        showInstruction: false,
                        field: "q5",
                        options: [
                            new KeyValuePair(this.XSTRING("q5_a2"), 2),
                            new KeyValuePair(this.XSTRING("q5_a1"), 1),
                            new KeyValuePair(this.XSTRING("q5_a0"), 0)
                        ]
                    }
                ]
            },

            // ----------------------------------------------------------------
            // Q6
            // ----------------------------------------------------------------
            {
                title: this.XSTRING('q6_title'),
                clinician: true,
                elements: [
                    {
                        type: "QuestionText",
                        bold: true,
                        text: this.XSTRING("q6_instruction_1")
                    },
                    {
                        type: "QuestionText",
                        text: this.XSTRING("q6_instruction_2")
                    },
                    {
                        type: "QuestionBooleanText",
                        mandatory: true,
                        text: this.XSTRING("q6_seq1"),
                        field: "q6_seq1"
                    },
                    {
                        type: "QuestionBooleanText",
                        mandatory: true,
                        text: this.XSTRING("q6_seq2"),
                        field: "q6_seq2"
                    },
                    {
                        type: "QuestionBooleanText",
                        mandatory: true,
                        text: this.XSTRING("q6_seq3"),
                        field: "q6_seq3"
                    },
                    {
                        type: "QuestionBooleanText",
                        mandatory: true,
                        text: this.XSTRING("q6_seq4"),
                        field: "q6_seq4"
                    },
                    {
                        // A simple QuestionImage expands beyond the right-
                        // hand edge. It would probably be worth fixing this.
                        // However, the simple quick fix is to put it inside
                        // a 1-column tabe, and then everything works fine.
                        type: "ContainerTable",
                        columns: 1,
                        elements: [
                            {
                                type: "QuestionImage",
                                image: IMAGE_SWM
                            }
                        ]
                    }
                ]
            },

            // ----------------------------------------------------------------
            // Q7
            // ----------------------------------------------------------------
            {
                title: this.XSTRING('q7_title'),
                clinician: true,
                elements: [
                    {
                        type: "QuestionText",
                        bold: true,
                        text: this.XSTRING("q7_proverb1")
                    },
                    {
                        type: "QuestionMCQ",
                        mandatory: true,
                        horizontal: false,
                        showInstruction: false,
                        field: "q7_proverb1",
                        options: proverb_options
                    },
                    {
                        type: "QuestionText",
                        bold: true,
                        text: this.XSTRING("q7_proverb2")
                    },
                    {
                        type: "QuestionMCQ",
                        mandatory: true,
                        horizontal: false,
                        showInstruction: false,
                        field: "q7_proverb2",
                        options: proverb_options
                    },
                    {
                        type: "QuestionText",
                        bold: true,
                        text: this.XSTRING("q7_proverb3")
                    },
                    {
                        type: "QuestionMCQ",
                        mandatory: true,
                        horizontal: false,
                        showInstruction: false,
                        field: "q7_proverb3",
                        options: proverb_options
                    }
                ]
            },

            // ----------------------------------------------------------------
            // Q8
            // ----------------------------------------------------------------
            {
                title: this.XSTRING('q8_title'),
                clinician: true,
                elements: [
                    {
                        type: "QuestionHeading",
                        bold: true,
                        text: this.XSTRING("q8_instruction_1")
                    },
                    {
                        type: "QuestionText",
                        bold: true,
                        text: this.XSTRING("q8_instruction_2")
                    },
                    {
                        type: "QuestionText",
                        bold: true,
                        text: this.XSTRING("q8_instruction_3")
                    },
                    {
                        type: "QuestionText",
                        text: this.XSTRING("q8_instruction_4")
                    },
                    {
                        type: "QuestionText",
                        bold: true,
                        text: this.XSTRING("q8_instruction_5")
                    },
                    {
                        type: "QuestionText",
                        text: this.XSTRING("q8_instruction_6")
                    },
                    {
                        type: "QuestionHeading",
                        bold: true,
                        text: this.XSTRING("q8_instruction_7")
                    },
                    {
                        type: "QuestionText",
                        bold: true,
                        text: this.XSTRING("q8_instruction_8")
                    },
                    {
                        type: "QuestionText",
                        bold: true,
                        text: this.XSTRING("q8_instruction_9")
                    },
                    { type: "QuestionHorizontalRule" },
                    {
                        type: "QuestionText",
                        bold: true,
                        text: this.XSTRING("q8_sentence_1")
                    },
                    {
                        type: "QuestionMCQ",
                        mandatory: true,
                        horizontal: false,
                        showInstruction: false,
                        field: "q8_sentence1",
                        options: inhibition_options
                    },
                    {
                        type: "QuestionText",
                        bold: true,
                        text: this.XSTRING("q8_sentence_2")
                    },
                    {
                        type: "QuestionMCQ",
                        mandatory: true,
                        horizontal: false,
                        showInstruction: false,
                        field: "q8_sentence2",
                        options: inhibition_options
                    },
                    {
                        type: "QuestionText",
                        bold: true,
                        text: this.XSTRING("q8_sentence_3")
                    },
                    {
                        type: "QuestionMCQ",
                        mandatory: true,
                        horizontal: false,
                        showInstruction: false,
                        field: "q8_sentence3",
                        options: inhibition_options
                    }
                ]
            }
        ];

        questionnaire = new Questionnaire({
            readOnly: readOnly,
            pages: pages,
            callbackThis: self,
            /* jshint unused:true */
            fnMakePageOnTheFly: function (pageId, pageTag) {
                if (pageTag !== Q4_PAGETAG) {
                    Titanium.API.error("fnMakePageOnTheFly: bad call");
                    return;
                }
                var elements = [
                        {
                            type: "QuestionText",
                            text: self.XSTRING("q4_instruction_1")
                        }
                    ],
                    seqlen,
                    pair,
                    required = true,
                    val1,
                    val2;
                for (seqlen = 2; seqlen <= 7; ++seqlen) {
                    for (pair = 1; pair <= 2; ++pair) {
                        elements.push({
                            elementTag: DIGIT_ELEMENT_PREFIX + seqlen,
                            type: "QuestionBooleanText",
                            mandatory: required,
                            visible: required,
                            text: self.XSTRING("q4_seq_len" + seqlen +
                                               "_" + pair),
                            field: "q4_len" + seqlen + "_" + pair
                        });
                    }
                    val1 = self["q4_len" + seqlen + "_1"];
                    val2 = self["q4_len" + seqlen + "_2"];
                    if (!val1 && !val2) {
                        // ... integers are returned for boolean fields;
                        // see BooleanWidget.toggle(). So don't use
                        // "=== false".
                        required = false;  // for subsequent items
                    }
                }
                return {
                    title: self.XSTRING('q4_title'),
                    clinician: true,
                    elements: elements
                };
            },
            fnGetFieldValue: self.defaultGetFieldValueFn,
            fnSetField: function (field, value) {
                var required,
                    seqlen,
                    tag,
                    val1,
                    val2;
                self.defaultSetFieldFn(field, value);
                if (lang.startsWith(field, "q4")) {
                    required = true;
                    for (seqlen = 2; seqlen <= 7; ++seqlen) {
                        tag = DIGIT_ELEMENT_PREFIX + seqlen;
                        questionnaire.setMandatoryByTag(tag, required);
                        questionnaire.setVisibleByTag(tag, required);
                        val1 = this["q4_len" + seqlen + "_1"];
                        val2 = this["q4_len" + seqlen + "_2"];
                        if (!val1 && !val2) {
                            // ... integers are returned for boolean fields;
                            // see BooleanWidget.toggle(). So don't use
                            // "=== false".
                            required = false;  // for subsequent items
                        }
                    }
                }
            },
            fnFinished: function (result, editing_time_s) {
                self.defaultFinishedFn(result, editing_time_s);
                questionnaire = null; // for garbage collection, since we have
                // closures referring to questionnaire
            }
        });
        questionnaire.open();
    }
});

module.exports = Ifs;
